### PR TITLE
feat: add dynamic len back references

### PR DESCRIPTION
**Pull request by gbotrel**
Improves compression ratio.

Rough idea; instead of having a fix number of bits to store the address of the backref, make that a function of our position in the stream. For example, if we are processing byte 278, addressable space is 277bits so we just need 9 bits to store the backref address. We have an arbitrary max bound size at 1 << 20 (20 bits). 

This PR also simplifies a bit the compression loop; using a circular buffer to avoid re-computing back references that we may have computed already as part of the previous "retry window i+1 i+2..." .

### CODE REVIEW

File: lzss/compress.go
Comment (Tabaie) [inline lzss/compress.go]: bits or bytes?

Comment (Tabaie) [reply lzss/compress.go]: I don't understand how the max length can be 8 bits, even if we're only talking about the address field.

**Review by Tabaie — COMMENTED** [review]
This works, but I suggest holding off on merging until after the release for two reasons:
1. Involves nontrivial circuit modifications that don't save on constraints
2. I think its "non-uniform" nature (varying backref length) can make things difficult should we choose to wizardify.

File: lzss/compress.go
Comment (gbotrel) [reply lzss/compress.go]: yes indeed 1 << xx; so in bytes 🙄

**Comment by gbotrel** [comment]
> This works, but I suggest holding off on merging until after the release for two reasons:
> 
> 1. Involves nontrivial circuit modifications that don't save on constraints
> 2. I think its "non-uniform" nature (varying backref length) can make things difficult should we choose to wizardify.

So counter arguments here; 
1. I thought modifications for that were trivial? (i.e. known at compile time / easy to implement?). If not indeed, no rush to merge, BUT
2. If we don't wizardify, which is also a likely scenario, we want to avoid breaking changes in the circuit, and... since we have a bunch of these breaking changes in the upcoming release, I'ld rather pack what we know we will do there rather than force a new circuit setup / blob parsing etc breaking change in the next one.

**Comment by Tabaie** [comment]
> > This works, but I suggest holding off on merging until after the release for two reasons:
> > 
> > 1. Involves nontrivial circuit modifications that don't save on constraints
> > 2. I think its "non-uniform" nature (varying backref length) can make things difficult should we choose to wizardify.
> 
> So counter arguments here;
> 
> 1. I thought modifications for that were trivial? (i.e. known at compile time / easy to implement?). If not indeed, no rush to merge, BUT
> 2. If we don't wizardify, which is also a likely scenario, we want to avoid breaking changes in the circuit, and... since we have a bunch of these breaking changes in the upcoming release, I'ld rather pack what we know we will do there rather than force a new circuit setup / blob parsing etc breaking change in the next one.

Okay I'm convinced once staticcheck is happy I'll merge.

**Review by Tabaie — APPROVED** [review]

**Pull request merged**
